src/mpid/ch4/generic/mpidig_send.h

```diff
@@ -14,20 +14,64 @@
 
 #include "ch4_impl.h"
 
+#define MPIDIG_AM_SEND_HDR_SIZE  sizeof(MPIDIG_hdr_t)
+
+static inline int mpidig_eager_limit(int is_local)
+{
+#ifdef MPIDI_CH4_DIRECT_NETMOD
+    return MPIDI_NM_am_eager_limit() - MPIDIG_AM_SEND_HDR_SIZE;
+#else
+    if (is_local) {
+        return MPIDI_SHM_am_eager_limit() - MPIDIG_AM_SEND_HDR_SIZE;
+    } else {
+        return MPIDI_NM_am_eager_limit() - MPIDIG_AM_SEND_HDR_SIZE;
+    }
+#endif
+}
+
+static inline int MPIDIG_do_send(const void *buf, MPI_Aint count, MPI_Datatype datatype,
+                                 int rank, int tag, MPIR_Comm * comm, int context_offset,
+                                 MPIDI_av_entry_t * addr, MPIR_Request ** request,
+                                 MPIR_Errflag_t errflag);
+static inline int MPIDIG_do_long(const void *buf, MPI_Aint count, MPI_Datatype datatype,
+                                 MPI_Aint data_sz,
+                                 int rank, int tag, MPIR_Comm * comm, int context_offset,
+                                 MPIDI_av_entry_t * addr, MPIR_Request ** request,
+                                 MPIR_Errflag_t errflag);
+static inline int MPIDIG_do_ssend(const void *buf, MPI_Aint count, MPI_Datatype datatype,
+                                  int rank, int tag, MPIR_Comm * comm, int context_offset,
+                                  MPIDI_av_entry_t * addr, MPIR_Request ** request,
+                                  MPIR_Errflag_t errflag);
+
 static inline int MPIDIG_isend_impl(const void *buf, MPI_Aint count, MPI_Datatype datatype,
                                     int rank, int tag, MPIR_Comm * comm, int context_offset,
                                     MPIDI_av_entry_t * addr, MPIR_Request ** request,
                                     int type, MPIR_Errflag_t errflag)
 {
-    int mpi_errno = MPI_SUCCESS, c;
-    MPIR_Request *sreq = *request;
-    MPIDIG_hdr_t am_hdr;
-    MPIDIG_ssend_req_msg_t ssend_req;
-    void *hdr = &am_hdr;
-    size_t hdr_sz = sizeof(am_hdr);;
+    MPI_Aint data_sz;
+    MPIR_Datatype *dt_ptr;
+    MPIDI_Datatype_get_size_dt_ptr(count, datatype, data_sz, dt_ptr);
 
-    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDIG_ISEND_IMPL);
-    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDIG_ISEND_IMPL);
+    int is_local = MPIDI_av_is_local(addr);
+    if (type == MPIDIG_SSEND_REQ) {
```

Comment:
No rndv for long synchronous sends? ch3 only implemented an eager ssend protocol. Synchronous sends larger than the eager limit went over the regular rndv protocol with no modification.

Reply:
That is how current ch4 code does. I am trying to stick to a refactor for this PR. RNDZ synchronous sends need be implemented as a separate protocol. If we factor the common code smartly, a new protocol can be implemented easily and cleanly.

Reply:
True, fair enough.